File: app/schemas/trace.py
from typing import Any

from pydantic import BaseModel, Field


TracePayload = dict[str, Any] | list[Any] | str | int | float | bool | None


class AgentTraceEventResponse(BaseModel):
    id: int
    trace_id: str
    run_id: str
    user_id: str | None = None
    session_id: str | None = None
    event_type: str
    step_index: int | None = None
    status_before: str | None = None
    status_after: str | None = None
    tool_name: str | None = None
    latency_ms: int | None = None
    input_summary: TracePayload = None
    output_summary: TracePayload = None
    error_code: str | None = None
    message: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: str


class AgentTraceDetailResponse(BaseModel):
    trace_id: str
    run_id: str | None = None
    user_id: str | None = None
    session_id: str | None = None
    event_count: int
    events: list[AgentTraceEventResponse] = Field(default_factory=list)

File: frontend/src/types/trace.ts
import type { JsonValue } from './agent'

export type TracePayload = JsonValue | JsonValue[] | Record<string, JsonValue>

export type AgentTraceEvent = {
  id: number
  trace_id: string
  run_id: string
  user_id?: string | null
  session_id?: string | null
  event_type: string
  step_index?: number | null
  status_before?: string | null
  status_after?: string | null
  tool_name?: string | null
  latency_ms?: number | null
  input_summary?: TracePayload | null
  output_summary?: TracePayload | null
  error_code?: string | null
  message?: string | null
  metadata: Record<string, JsonValue>
  created_at: string
}

export type AgentTraceDetailResponse = {
  trace_id: string
  run_id?: string | null
  user_id?: string | null
  session_id?: string | null
  event_count: number
  events: AgentTraceEvent[]
}
